components/p3a/component_installer.cc

```diff
@@ -0,0 +1,112 @@
+#include "brave/components/brave_component_updater/browser/brave_on_demand_updater.h"
+#include "brave/components/p3a/remote_config_manager.h"
+#include "components/component_updater/component_installer.h"
+#include "components/component_updater/component_updater_service.h"
+
+using component_updater::ComponentUpdateService;
+
+namespace p3a {
+
+namespace {
+
+constexpr char kP3AComponentName[] = "P3A Configuration";
+constexpr char kP3AComponentId[] = "memmkfnkoigleebghlpgeiecaddoblcl";
+constexpr base::FilePath::CharType kComponentInstallDir[] =
+    FILE_PATH_LITERAL("P3AConfig");
+// public key hash:
+// c4cca5dae86b44167bf64842033e1b2b5c908bd6abc30e32d08b12bf50b2cc03
+const uint8_t kP3AComponentHash[32] = {
+    0xc4, 0xcc, 0xa5, 0xda, 0xe8, 0x6b, 0x44, 0x16, 0x7b, 0xf6, 0x48,
+    0x42, 0x03, 0x3e, 0x1b, 0x2b, 0x5c, 0x90, 0x8b, 0xd6, 0xab, 0xc3,
+    0x0e, 0x32, 0xd0, 0x8b, 0x12, 0xbf, 0x50, 0xb2, 0xcc, 0x03};
+
+class P3AComponentInstallerPolicy
+    : public component_updater::ComponentInstallerPolicy {
+ public:
+  explicit P3AComponentInstallerPolicy(
+      base::WeakPtr<RemoteConfigManager> remote_config_manager)
+      : remote_config_manager_(remote_config_manager) {}
+  ~P3AComponentInstallerPolicy() override = default;
+
+  P3AComponentInstallerPolicy(const P3AComponentInstallerPolicy&) = delete;
+  P3AComponentInstallerPolicy& operator=(const P3AComponentInstallerPolicy&) =
+      delete;
+
+ private:
+  // component_updater::ComponentInstallerPolicy:
+  bool SupportsGroupPolicyEnabledComponentUpdates() const override {
+    return false;
+  }
+
+  bool RequiresNetworkEncryption() const override { return false; }
+
+  update_client::CrxInstaller::Result OnCustomInstall(
+      const base::Value::Dict& manifest,
+      const base::FilePath& install_dir) override {
+    return update_client::CrxInstaller::Result(0);
+  }
+
+  void OnCustomUninstall() override {}
+
+  bool VerifyInstallation(const base::Value::Dict& manifest,
+                          const base::FilePath& install_dir) const override {
+    return base::PathExists(install_dir.Append(kP3AManifestFileName));
+  }
+
+  void ComponentReady(const base::Version& version,
+                      const base::FilePath& install_dir,
+                      base::Value::Dict manifest) override {
+    remote_config_manager_->OnComponentReady(install_dir);
+  }
+
+  base::FilePath GetRelativeInstallDir() const override {
+    return base::FilePath(kComponentInstallDir);
+  }
+
+  void GetHash(std::vector<uint8_t>* hash) const override {
+    hash->assign(std::begin(kP3AComponentHash), std::end(kP3AComponentHash));
+  }
+
+  std::string GetName() const override { return kP3AComponentName; }
+
+  update_client::InstallerAttributes GetInstallerAttributes() const override {
+    return update_client::InstallerAttributes();
+  }
+
+  bool IsBraveComponent() const override { return true; }
+
+  base::WeakPtr<RemoteConfigManager> remote_config_manager_;
```

Comment:
why do we store a weak pointer but never check it? perhaps we shouldn't just instantiate the policy without alive config manager?

Reply:
Are you asking why we don't check the pointer for validity, or why we don't use the pointer? We check the pointer in `RegisterP3AComponent`, and use the pointer in `ComponentReady`.

Reply:
I mean, checking the pointer. If we already ensured that the pointer is non-null when it is passed to `P3AComponentInstallerPolicy` constructor, why to keep the weak pointer at all? In case you don't check it before using in ComponentReady

Reply:
Basically it is the thing that you are assumed to check for validity before using, otherwise you probably don't need a weak pointer

Reply:
added pointer check